// Code generated by mockery. DO NOT EDIT.

package billingenginemocks

import (
	context "context"

	usecases "github.com/JoshuaPangaribuan/billing-engine/internal/billing-engine/internal/usecases"
	mock "github.com/stretchr/testify/mock"
)

// MockGetOutstandingUsecase is an autogenerated mock type for the GetOutstandingUsecase type
type MockGetOutstandingUsecase struct {
	mock.Mock
}

type MockGetOutstandingUsecase_Expecter struct {
	mock *mock.Mock
}

func (_m *MockGetOutstandingUsecase) EXPECT() *MockGetOutstandingUsecase_Expecter {
	return &MockGetOutstandingUsecase_Expecter{mock: &_m.Mock}
}

// Execute provides a mock function with given fields: ctx, customerID, loanID
func (_m *MockGetOutstandingUsecase) Execute(ctx context.Context, customerID uint64, loanID uint64) (usecases.GetOutstandingOutput, error) {
	ret := _m.Called(ctx, customerID, loanID)

	if len(ret) == 0 {
		panic("no return value specified for Execute")
	}

	var r0 usecases.GetOutstandingOutput
	var r1 error
	if rf, ok := ret.Get(0).(func(context.Context, uint64, uint64) (usecases.GetOutstandingOutput, error)); ok {
		return rf(ctx, customerID, loanID)
	}
	if rf, ok := ret.Get(0).(func(context.Context, uint64, uint64) usecases.GetOutstandingOutput); ok {
		r0 = rf(ctx, customerID, loanID)
	} else {
		r0 = ret.Get(0).(usecases.GetOutstandingOutput)
	}

	if rf, ok := ret.Get(1).(func(context.Context, uint64, uint64) error); ok {
		r1 = rf(ctx, customerID, loanID)
	} else {
		r1 = ret.Error(1)
	}

	return r0, r1
}

// MockGetOutstandingUsecase_Execute_Call is a *mock.Call that shadows Run/Return methods with type explicit version for method 'Execute'
type MockGetOutstandingUsecase_Execute_Call struct {
	*mock.Call
}

// Execute is a helper method to define mock.On call
//   - ctx context.Context
//   - customerID uint64
//   - loanID uint64
func (_e *MockGetOutstandingUsecase_Expecter) Execute(ctx interface{}, customerID interface{}, loanID interface{}) *MockGetOutstandingUsecase_Execute_Call {
	return &MockGetOutstandingUsecase_Execute_Call{Call: _e.mock.On("Execute", ctx, customerID, loanID)}
}

func (_c *MockGetOutstandingUsecase_Execute_Call) Run(run func(ctx context.Context, customerID uint64, loanID uint64)) *MockGetOutstandingUsecase_Execute_Call {
	_c.Call.Run(func(args mock.Arguments) {
		run(args[0].(context.Context), args[1].(uint64), args[2].(uint64))
	})
	return _c
}

func (_c *MockGetOutstandingUsecase_Execute_Call) Return(_a0 usecases.GetOutstandingOutput, _a1 error) *MockGetOutstandingUsecase_Execute_Call {
	_c.Call.Return(_a0, _a1)
	return _c
}

func (_c *MockGetOutstandingUsecase_Execute_Call) RunAndReturn(run func(context.Context, uint64, uint64) (usecases.GetOutstandingOutput, error)) *MockGetOutstandingUsecase_Execute_Call {
	_c.Call.Return(run)
	return _c
}

// NewMockGetOutstandingUsecase creates a new instance of MockGetOutstandingUsecase. It also registers a testing interface on the mock and a cleanup function to assert the mocks expectations.
// The first argument is typically a *testing.T value.
func NewMockGetOutstandingUsecase(t interface {
	mock.TestingT
	Cleanup(func())
}) *MockGetOutstandingUsecase {
	mock := &MockGetOutstandingUsecase{}
	mock.Mock.Test(t)

	t.Cleanup(func() { mock.AssertExpectations(t) })

	return mock
}
